src/mail.sh

```diff
@@ -0,0 +1,289 @@
+# this will warn the user and clear the option.
+#
+# @option: The option to determine if it should be an email
+# @value:  The value being passed
+#
+# Return:
+# Returns 0 if valid; 22 if invalid
+function validate_email()
+{
+  local option="$1"
+  local value="$2"
+
+  if [[ "$option" =~ ^(email|smtpuser)$ ]]; then
+    if [[ ! "$value" =~ ^[A-Za-z0-9_\.-]+@[A-Za-z0-9_-]+(\.[A-Za-z0-9]+)+$ ]]; then
+      complain "Invalid $option: $value"
+      return 22 #EINVAL
+    fi
+  fi
+
+  return 0
+}
+
+# This function adds the supplied configuration option. It first checks if the
+# given option has been set, then prompts the user to replace, add, or ignore
+# the configuration
+#
+# @option:     The option being edited
+# @value:      The value to update the option to
+# @curr_scope: The scope being edited
+# @cmd_scope:  The scope being imposed on the commands
+# @set_option: The relevant index to access set_confs
+#
+# Return:
+# Returns 0 if successful; non-zero otherwise
+function add_config()
+{
+  local flag="$1"
+  local option="$2"
+  local value="${options_values[$option]}"
+  local curr_scope=${options_values['SCOPE']}
+  local cmd_scope=${options_values['CMD_SCOPE']}
+  local set_option="${curr_scope}_$option"
+  local cmd
+
+  flag=${flag:-'SILENT'}
+
+  if [[ "${options_values['FORCE']}" == 0 ]]; then
+    if [[ -n "${set_confs[$set_option]}" ]]; then
+      warning "The configuration $option is already set with the following value(s):"
+      for scp in {'global','local'}; do
+        if [[ -n "${set_confs[${scp}_$option]}" ]]; then
+          warning -n "  [$scp]: "
+          printf '%s\n' "${set_confs[${scp}_$option]}"
+        fi
+      done
+
+      printf '%s\n' '' "You are currently editing at the [$curr_scope] scope." \
+        'If you continue the value at this scope will be overwritten.' \
+        "The new value will be '$value'"
+
+      if [[ $(ask_yN 'Do you wish to proceed?') == 0 ]]; then
+        complain "Skipping $option..."
+        return 125 # ECANCELED
+      fi
+    fi
+  fi
+
+  cmd="git config --$cmd_scope $option '$value'"
+
+  cmd_manager "$flag" "$cmd"
+
+  if [[ "$?" == 0 ]] && [[ "$flag" != 'TEST_MODE' ]]; then
+    success "$option was successfully set to '${options_values[$option]}'"
+  fi
+}
+
+# This function gets all the currently set values for the configurations used
+# by this script and writes them to the global variable set_confs
+#
+# @cmd_scope:  The scope being imposed on the command
+# @set_values: The values of the option that are already set
+# @option:     The option name
+# @value:      The value to update the option to
+# @scope:      The scope of the given value
+#
+function get_configs()
+{
+  local cmd_scope=${options_values['CMD_SCOPE']}
+  local -a set_values
+  local option
+  local value
+  local scope
+  local i=0
+
+  IFS=$'\t\n' read -r -a set_values -d '' <<< "$(get_git_config_regex 'user|sendemail\.smtp' "$cmd_scope")"
```

Comment:
iirc, the safe way to deal with `IFS` is declaring it as a local.

Reply:
When used with the read command like this the `IFS` variable is only changed for the duration of the command. I have tested and can confirm that the value of the `IFS` variable was the same before and after the execution of this line.